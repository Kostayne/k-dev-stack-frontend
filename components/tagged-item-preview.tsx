import Link from 'next/link';
import React from 'react';
import * as RM from 'react-modifier';

export interface TaggedItemPreviewProps {
    headMod?: RM.IModifier;
    name: string;
    description: string;
    tags: string[];
    href: string;
}

const TaggedItemPreview= (props: TaggedItemPreviewProps) => {
    const headMod = props.headMod || RM.createMod();

    const getTagsToR = () => {
        return props.tags.map((t, i) => {
            return (
                <span key={i} className="text-robotoCond"># {t}</span>
            );
        });
    };

    return (
        RM.modElement((
            <div className="shadow-preview rounded-[4px] p-4 min-h-[180px]">
                <Link href={props.href}>
                    <a className="text-roboto font-medium text-contrastAlt">
                        {props.name}
                    </a>
                </Link>

                <p className="mt-[11px] text-robotoCond">{props.description}</p>

                <div className="flex flex-wrap gap-2 mt-[10px]">
                    {getTagsToR()}
                </div>
            </div>
        ), headMod)
    );
};

export default TaggedItemPreview;