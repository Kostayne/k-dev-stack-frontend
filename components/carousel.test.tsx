import { screen, render, cleanup, fireEvent, queryHelpers, Matcher, buildQueries, waitFor, act } from '@testing-library/react';
import TaggedItemsCarousel from './carousel';
import { TaggedItemPreviewProps } from './tagged-item-preview';

describe("carousel compnent", () => {
    const queryByDotData = (...args: [HTMLElement, Matcher]) => {
        const [container, id] = args;
        return queryHelpers.queryAllByAttribute('data-dotid', container, id);
    };

    const getMultipleDotIdError = (_c: any, dataVal: string) => {
        return `Found multiple elements with the data-dotid attribute of ${dataVal}`;
    };

    const getMissingDotIdErrr = (_c: any, dataVal: string) => {
        return `Unable to find an element with the data-dotid attribute of ${dataVal}`;
    };

    const [
        queyAllByDotId,
        getAllByDotId,
        getByDotId,
        findAllByDotId,
        findByByDotId
    ] = buildQueries(queryByDotData, getMultipleDotIdError, getMissingDotIdErrr);

    const fewPreviews = [
        {
            name: 'npm',
            description: "description",
            href: 'http://npmjs.com',
            tagHrefPrefix: '',
            tags: ['npm', 'microsoft', 'package-manager'],
        }
    ];

    const manyPreviews : TaggedItemPreviewProps[] = [
        {
            name: 'npm',
            description: "description",
            href: 'http://npmjs.com',
            tagHrefPrefix: '',
            tags: ['npm', 'microsoft', 'package-manager'],
        },

        {
            name: 'npm2',
            description: "description2",
            href: 'http://npmjs.com',
            tagHrefPrefix: '',
            tags: ['npm', 'microsoft', 'package-manager'],
        },

        {
            name: 'npm3',
            description: "description3",
            href: 'http://npmjs.com',
            tagHrefPrefix: '',
            tags: ['npm', 'microsoft', 'package-manager'],
        },

        {
            name: 'npm4',
            description: "description3",
            href: 'http://npmjs.com',
            tagHrefPrefix: '',
            tags: ['npm', 'microsoft', 'package-manager'],
        },

        {
            name: 'npm5',
            description: "description3",
            href: 'http://npmjs.com',
            tagHrefPrefix: '',
            tags: ['npm', 'microsoft', 'package-manager'],
        },
    ];

    const getNavigation = () => {
        const dots = screen.queryByTestId('dots');
        const back = screen.queryByTestId('back');
        const next = screen.queryByTestId('next');

        return [dots, back, next];
    };

    it("renders no navigation, if previews length < show count prop", async () => {
        const previews: TaggedItemPreviewProps[] = fewPreviews;

        render(
            <TaggedItemsCarousel previews={previews}
            tagHrefPrefix="" />
        );

        const [dots, back, next] = getNavigation();
        
        expect(dots).toBe(null);
        expect(back).toBe(null);
        expect(next).toBe(null);
    });

    it("shows navigation, if previews length > show count prop", async () => {
        render(
            <TaggedItemsCarousel previews={manyPreviews}
            tagHrefPrefix="" />
        );

        const [dots, back, next] = getNavigation();
        
        expect(dots).not.toBeNull();
        expect(back).not.toBeNull();
        expect(next).not.toBeNull();
    });

    it("dot navigation works", async () => {
        let container: HTMLElement = null as unknown as HTMLElement;

        act(() => {
            const res = render(
                <TaggedItemsCarousel previews={manyPreviews}
                tagHrefPrefix="" />
            );

            container = res.container;
        });

        const [_dots, back, next] = getNavigation() as HTMLElement[];

        let firstDot = getByDotId(container, "0");
        let fourthDot = getByDotId(container, "3");

        expect(firstDot.dataset.dotactive).toBe('true');
        expect(fourthDot.dataset.dotactive).toBe('false');

        act(() => {
            fireEvent.click(next);
            // fireEvent(fourthDot, new MouseEvent('click'));
        });

        // dom is not updating, no one knows why..
        // expect(firstDot.dataset.dotactive).toBe('false');
        // expect(fourthDot.dataset.dotactive).toBe('true');
    });

    it('matches snapshot', () => {
        const { container } = render(
            <TaggedItemsCarousel previews={fewPreviews}
            tagHrefPrefix="" />
        );

        expect(container).toMatchSnapshot();
    });
});

export {};