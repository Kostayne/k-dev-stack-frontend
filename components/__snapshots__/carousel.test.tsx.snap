// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`carousel compnent matches snapshot 1`] = `
<div>
  <div
    class="carousel"
  >
    <div
      class=""
    >
      <div
        aria-label="slider"
        aria-live="polite"
        class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal max-h-[200px] min-h-[200px] undefined"
        role="listbox"
        tabindex="0"
      >
        <div
          class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
        >
          <div
            class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
            style="width: 100%; transform: translateX(0%) translateX(0px); flex-direction: row;"
          >
            <div
              aria-label="slide"
              aria-selected="true"
              class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
              role="option"
              style="width: 33.333333333333336%; padding-bottom: 4.640371229698376%;"
              tabindex="0"
            >
              <div
                class="slideInner___2mfX9 carousel__inner-slide !pr-[15px]"
              >
                <div
                  class="flex flex-col shadow-baseShadow rounded-[4px] p-4 min-h-[180px]"
                >
                  <a
                    class="text-roboto font-medium text-contrastAlt"
                    href="http://npmjs.com"
                  >
                    npm
                  </a>
                  <p
                    class="mt-[11px] text-robotoCond"
                  >
                    description
                  </p>
                  <div
                    class="flex flex-wrap gap-2 mt-[auto] pt-[10px]"
                  >
                    <a
                      class="text-robotoCond text-contrast"
                      href="/npm"
                    >
                      # 
                      npm
                    </a>
                    <a
                      class="text-robotoCond text-contrast"
                      href="/microsoft"
                    >
                      # 
                      microsoft
                    </a>
                    <a
                      class="text-robotoCond text-contrast"
                      href="/package-manager"
                    >
                      # 
                      package-manager
                    </a>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-label="slide"
              aria-selected="true"
              class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
              role="option"
              style="width: 33.333333333333336%; padding-bottom: 4.640371229698376%;"
              tabindex="0"
            >
              <div
                class="slideInner___2mfX9 carousel__inner-slide !pr-[15px]"
              >
                <div
                  class="flex flex-col shadow-baseShadow rounded-[4px] p-4 min-h-[180px]"
                >
                  <span
                    class="text-roboto font-medium text-customGray"
                  >
                    Не существеут
                  </span>
                  <p
                    class="mt-[11px] text-robotoCond"
                  >
                    Данной вещи пока нет на сайте.. возможно, скоро она появится.
                  </p>
                </div>
              </div>
            </div>
            <div
              aria-label="slide"
              aria-selected="true"
              class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
              role="option"
              style="width: 33.333333333333336%; padding-bottom: 4.640371229698376%;"
              tabindex="0"
            >
              <div
                class="slideInner___2mfX9 carousel__inner-slide !pr-[15px]"
              >
                <div
                  class="flex flex-col shadow-baseShadow rounded-[4px] p-4 min-h-[180px]"
                >
                  <span
                    class="text-roboto font-medium text-customGray"
                  >
                    Не существеут
                  </span>
                  <p
                    class="mt-[11px] text-robotoCond"
                  >
                    Данной вещи пока нет на сайте.. возможно, скоро она появится.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
